document arrays in sql (#12549)
apache/druid: docs/querying/sql-data-types.md

> Druid natively supports five basic column types: "long" (64 bit signed int), "float" (32 bit float), "double" (64 bit
> float) "string" (UTF-8 encoded strings and string arrays), and "complex" (catch-all for more exotic data types like

Reviewer: ```suggestion
Druid natively supports the following basic column types: "long" (64-bit signed int), "float" (32-bit float), "double" (64-bit
float) "string" (UTF-8 encoded strings and string arrays), "complex" (catch-all for more exotic data types like
```

Reviewer 2: ```suggestion
Druid natively supports the following column types: "long" (64-bit signed int), "float" (32-bit float), "double" (64-bit
float) "string" (UTF-8 encoded strings and string arrays), and "complex" (non-standard data types, such as nested JSON, hyperUnique and approxHistogram aggregators, and DataSketches aggregators)
```

Reviewer: I like the list formatting for readability; don't forget to remove L36 if using this option

|BIGINT|LONG|`0`|Druid LONG columns (except `__time`) are reported as BIGINT|
|TIMESTAMP|LONG|`0`, meaning 1970-01-01 00:00:00 UTC|Druid's `__time` column is reported as TIMESTAMP. Casts between string and timestamp types assume standard SQL formatting, e.g. `2000-01-02 03:04:05`, _not_ ISO8601 formatting. For handling other formats, use one of the [time functions](sql-scalar.md#date-and-time-functions).|
|DATE|LONG|`0`, meaning 1970-01-01|Casting TIMESTAMP to DATE rounds down the timestamp to the nearest day. Casts between string and date types assume standard SQL formatting, e.g. `2000-01-02`. For handling other formats, use one of the [time functions](sql-scalar.md#date-and-time-functions).|

> |ARRAY|ARRAY|Druid native array types work as SQL arrays, and multi-value strings can be converted to arrays. See the [`ARRAY` details](#arrays) for more details.|

Reviewer 2: ```suggestion
|ARRAY|ARRAY|Druid native array types work as SQL arrays, and multi-value strings can be converted to arrays. See [Arrays](#arrays) for more details.|
```

|BIGINT|LONG|`0`|Druid LONG columns (except `__time`) are reported as BIGINT|
|TIMESTAMP|LONG|`0`, meaning 1970-01-01 00:00:00 UTC|Druid's `__time` column is reported as TIMESTAMP. Casts between string and timestamp types assume standard SQL formatting, e.g. `2000-01-02 03:04:05`, _not_ ISO8601 formatting. For handling other formats, use one of the [time functions](sql-scalar.md#date-and-time-functions).|
|DATE|LONG|`0`, meaning 1970-01-01|Casting TIMESTAMP to DATE rounds down the timestamp to the nearest day. Casts between string and date types assume standard SQL formatting, e.g. `2000-01-02`. For handling other formats, use one of the [time functions](sql-scalar.md#date-and-time-functions).|

> |ARRAY|ARRAY|Druid native array types work as SQL arrays, and multi-value strings can be converted to arrays. See the [`ARRAY` details](#arrays).|

Reviewer: I think you're missing a column value. You have three columns but should have four. The third column should be default value, and the fourth column is notes.

> they are handled in Druid SQL and in native queries. For example, expressions involving multi-value dimensions may be
> incorrectly optimized by the Druid SQL planner: `multi_val_dim = 'a' AND multi_val_dim = 'b'` will be optimized to
> `false`, even though it is possible for a single row to have both "a" and "b" as values for `multi_val_dim`. The
> SQL behavior of multi-value dimensions may change in a future release to more closely align with their behavior
> in native queries, but the [multi-value string functions](./sql-multivalue-string-functions.md) should be able to provide
> nearly all possible native functionality.

## Arrays
Multi-value dimensions may also be converted to standard SQL arrays, either by explicitly converting them with `MV_TO_ARRAY`,

> or implicitly when used within the [array functions](./sql-array-functions.md). `ARRAY` types behave as standard SQL arrays, where

Reviewer 2: ```suggestion
or implicitly when used within the [array functions](./sql-array-functions.md). `ARRAY` types behave as standard SQL arrays.
```
Consider splitting this into two sentences.

> grouping on them will group on the entire array of values instead of the implicit `UNNEST` that occurs when grouping on

Reviewer 2: ```suggestion
Grouping on `ARRAY` types results in grouping on the entire array of values instead of the implicit `UNNEST` that occurs when grouping on
```

multi-value dimensions directly or when used with the multi-value functions. Arrays may also be constructed from multiple
columns using the array functions.

## Multi-value strings behavior
The behavior of Druid [multi-value string dimensions](multi-value-dimensions.md) varies depending on the context of their usage.

> When used as `VARCHAR` functions, which are not "aware" that their inputs which claim to be `VARCHAR` might actually have multiple

Reviewer 2: ```suggestion
When used as `VARCHAR` functions, whose inputs might have multiple
```

> values such as `CONCAT`, Druid will map the function across all values in the row. If the row is null or empty, the function will

Reviewer 2: ```suggestion
values such as `CONCAT`, Druid maps the function across all values in the row. If the row is null or empty, the function
```
Removed future tense.

> recieve `NULL` as its input, otherwise it will be applied to every row value and continue its life as a multi-value VARCHAR.

Reviewer 2: ```suggestion
recieve `NULL` as its input, otherwise it is applied to every row value and continues its life as a multi-value `VARCHAR`.
```

> When used with the explicit [multi-value string functions](./sql-multivalue-string-functions.md), the column is acknowledged to be multi-valued,

Reviewer 2: ```suggestion
When you apply [multi-value string functions](./sql-multivalue-string-functions.md) to a multi-valued column, Druid processes the column values as `ARRAY` typed. Operations that produce null and empty rows are distinguished as separate values (unlike implicit mapping behavior), but retain their `VARCHAR` type after the computation is complete.
```
Simplified the text to improve readability.

and during processing the values are operated on as if they were `ARRAY` typed, so any operations which produce null and empty rows are
distinguished as separate values (unlike implicit mapping behavior), but retain their `VARCHAR` type after the computation is complete.

> Note that Druid multi-value columns do _not_ distinguish between empty and null rows, so an empty row will never appear natively as input

Reviewer 2: ```suggestion
Note that Druid multi-value columns do not distinguish between empty and null rows. An empty row never appears natively as an input to a multi-valued function, but a multi-value function that manipulates the array form of the value may produce an empty array, which is handled separately while processing.
```

to a multi-valued function, but any multi-value function which manipulates the array form of the value may produce an empty array, which
will be handled separately while processing.

> Do not mix the usage of multi-value functions and normal scalar functions within the same expression, as the planner will be unable
> to determine how to properly process the value given its ambiguous usage. A multi-value string must be treated consistently within
> an expression.

> Finally, when converted to `ARRAY` or used with [array functions](./sql-array-functions.md), they behave as standard SQL arrays and can no longer

Reviewer 2: ```suggestion
When converted to `ARRAY` or used with [array functions](./sql-array-functions.md), multi-value string dimensions behave as standard SQL arrays and can no longer
```

be manipulated with non-array functions.

> When `VARCHAR` typed, multi-value results produced by non-aggregating selection of values will be serialized into a JSON string of the array.

Reviewer 2: ```suggestion
Druid serializes `VARCHAR` typed results that are produced by non-aggregating selection of values into a JSON string of the array.
```
Removed passive voice.

> `ARRAY` typed results will be serialized into stringified JSON arrays if the context parameter `sqlStringifyArrays` is set, otherwise they

Reviewer 2: ```suggestion
Druid serializes `ARRAY` typed results into stringified JSON arrays if the context parameter `sqlStringifyArrays` is set, otherwise they remain in their array format.
```
Removed passive voice.

> they are handled in Druid SQL and in native queries. For example, expressions involving multi-value dimensions may be
> incorrectly optimized by the Druid SQL planner: `multi_val_dim = 'a' AND multi_val_dim = 'b'` is optimized to
> `false`, even though it is possible for a single row to have both "a" and "b" as values for `multi_val_dim`. The
> SQL behavior of multi-value dimensions may change in a future release to more closely align with their behavior
> in native queries, but the [multi-value string functions](./sql-multivalue-string-functions.md) should be able to provide
> nearly all possible native functionality.

## Arrays
Druid supports `ARRAY` types constructed at query time, though it currently lacks the ability to store them in
segments. `ARRAY` types behave as standard SQL arrays, where results are grouped by matching entire arrays. This is in
contrast to the implicit `UNNEST` that occurs when grouping on multi-value dimensions directly or when used with the

> multi-value functions. You can convert multi-value dimensions to standard SQL arrays either by explicitly converting
> them with `MV_TO_ARRAY`, or implicitly when used within the [array functions](./sql-array-functions.md). Arrays may

Reviewer: ```suggestion
multi-value functions. You can convert multi-value dimensions to standard SQL arrays either explicitly by converting
them with `MV_TO_ARRAY` or implicitly when used within the [array functions](./sql-array-functions.md). Arrays may
```

also be constructed from multiple columns using the array functions.

## Multi-value strings behavior
The behavior of Druid [multi-value string dimensions](multi-value-dimensions.md) varies depending on the context of their usage.

When used with standard `VARCHAR` functions, which expect a single input value per row, Druid maps the function across all values in the row.

> the function across all values in the row. If the row is null or empty, the function will recieve `NULL` as its input,

Reviewer: ```suggestion
If the row is null or empty, the function will receive `NULL` as its input,
```

Reviewer 2: ```suggestion
the function across all values in the row. If the row is null or empty, the function receives `NULL` as its input,
```

Reviewer 2: Try to avoid future tense when possible.

otherwise it will be applied to every row value and continue its life as a multi-value `VARCHAR`.

When used with the explicit [multi-value string functions](./sql-multivalue-string-functions.md), Druid processes the
row values as if they were `ARRAY` typed, so any operations which produce null and empty rows are distinguished as
separate values (unlike implicit mapping behavior). These multi-value string functions, typically denoted with an `MV_`
prefix, retain their `VARCHAR` type after the computation is complete. Note that Druid multi-value columns do _not_

> distinguish between empty and null rows, so an empty row will never appear natively as input to a multi-valued function,

Reviewer: Separated the sentences for readability
```suggestion
distinguish between empty and null rows. An empty row will never appear natively as input to a multi-valued function,
```

> but any multi-value function which manipulates the array form of the value may produce an empty array, which will be

Reviewer: ```suggestion
but any multi-value function which manipulates the array form of the value may produce an empty array, which is
```

handled separately while processing.

> Do not mix the usage of multi-value functions and normal scalar functions within the same expression, as the planner will be unable
> to determine how to properly process the value given its ambiguous usage. A multi-value string must be treated consistently within
> an expression.

When converted to `ARRAY` or used with [array functions](./sql-array-functions.md), multi-value strings behave as standard SQL arrays and can no longer
be manipulated with non-array functions.

> Druid serializes multi-value `VARCHAR` results as a JSON string of the array, if the value was not grouped on. If the

Reviewer: ```suggestion
Druid serializes multi-value `VARCHAR` results as a JSON string of the array, if grouping was not applied on the value. If the
```

> nearly all possible native functionality.

## Arrays

> Druid also has support for `ARRAY` types constructed at query time, though currently lacks the ability to store them in

Reviewer 2: ```suggestion
Druid supports `ARRAY` types constructed at query time, though it currently lacks the ability to store them in
```

segments. `ARRAY` types behave as standard SQL arrays, where results are grouped by matching entire arrays. This is in
contrast to the implicit `UNNEST` that occurs when grouping on multi-value dimensions directly or when used with the
multi-value functions. You can convert multi-value dimensions to standard SQL arrays either by explicitly converting
them with `MV_TO_ARRAY`, or implicitly when used within the [array functions](./sql-array-functions.md). Arrays may
also be constructed from multiple columns using the array functions.

## Multi-value strings behavior
The behavior of Druid [multi-value string dimensions](multi-value-dimensions.md) varies depending on the context of their usage.

> When used with standard `VARCHAR` functions which expect a single input value per row, such as `CONCAT`, Druid will map

Reviewer 2: ```suggestion
When used with standard `VARCHAR` functions, which expect a single input value per row, Druid maps the function across all values in the row. If the row is null or empty, the function receives `NULL` as its input, otherwise it is applied to every row value and continues its life as a multi-value `VARCHAR`.
```
Do you think we can omit the `CONCAT` example in this case? I think the sentence reads better without it.
